src/Bicep.Core/Semantics/Namespaces/SystemNamespaceSymbol.cs

```diff
@@ -504,6 +505,56 @@ private static SyntaxBase StringLiteralFunctionReturnTypeEvaluator(FunctionCallS
             return SyntaxFactory.CreateStringLiteral(stringLiteral.RawStringValue);
         }
 
+        private static TypeSymbol JsonTypeBuilder(IBinder binder, IFileResolver fileResolver, IDiagnosticWriter diagnostics, ImmutableArray<FunctionArgumentSyntax> arguments, ImmutableArray<TypeSymbol> argumentTypes)
+        {
+            static TypeSymbol ToBicepType(JToken token)
+                => token switch {
+                    JObject @object => new ObjectType(
+                        "object",
+                        TypeSymbolValidationFlags.Default,
+                        @object.Properties().Select(x => new TypeProperty(x.Name, ToBicepType(x.Value), TypePropertyFlags.ReadOnly | TypePropertyFlags.ReadableAtDeployTime)),
+                        null),
+                    JArray @array => new TypedArrayType(
+                        UnionType.Create(@array.Select(x => ToBicepType(x))),
+                        TypeSymbolValidationFlags.Default),
+                    JValue value => value.Type switch {
+                        JTokenType.String => new StringLiteralType(value.ToString()),
+                        JTokenType.Integer => LanguageConstants.Int,
+                        // Floats are currently not supported in Bicep, so fall back to the default behavior of "any"
+                        JTokenType.Float => LanguageConstants.Any,
+                        JTokenType.Boolean => LanguageConstants.Bool,
+                        JTokenType.Null => LanguageConstants.Null,
+                        _ => LanguageConstants.Any,
+                    },
+                    _ => LanguageConstants.Any,
+                };
+
+            if (argumentTypes.Length != 1 || argumentTypes[0] is not StringLiteralType stringLiteral)
+            {
+                return LanguageConstants.Any;
+            }
+
+            try
+            {
+                var jsonLoadSettings = new JsonLoadSettings
+                {
+                    CommentHandling = CommentHandling.Ignore,
+                    DuplicatePropertyNameHandling = DuplicatePropertyNameHandling.Error,
+                    LineInfoHandling = LineInfoHandling.Ignore
+                };
+
+                var token = JToken.Parse(stringLiteral.RawStringValue, jsonLoadSettings);
+
+                return ToBicepType(token);
+            }
+            catch
+            {
+                var error = DiagnosticBuilder.ForPosition(arguments[0].Expression).UnparseableJsonType();
```

Comment:
>UnparseableJsonType [](http://example.com/codeflow?start=83&length=19)

Are the exception messages good enough to pass through to the diagnostic? It would provide additional info about what is wrong with the JSON.

Reply:
We could. Here are some examples:
* `After parsing a value an unexpected character was encountered: s. Path 'abc', line 5, position 0.`
* `Unterminated string. Expected delimiter: \". Path 'abc', line 7, position 1."`

Initiallly I was imagining we'd fall back to returning `any`, but I'd be more keen on returning info now that I'm returning an error. One concern is that the line info attached to the error may be confusing - especially if using e.g. line control characters like `r` `\n` in the Bicep string.

Reply:
Yeah that's a good pt. The line numbers will be relative to the value. Technically we could put the squiggles in that location, but that feels like adding too much complexity to this feature.

If the JSON is getting so complex that it's getting hard to figure out issues in it, our recommendation should be to switch to a separate file and pull it in via `loadTextContent()`. Then, the built-in VS code language server will provide a good troubleshooting experience.

Reply:
I agree - I've added a comment about this and switched to using ARM's .TryFromJson method.